improve: convert HTML currently in script editor error messages sent to screen readers

atari2600tim opened the pull request:
<!-- Keep the title short & concise so anyone non-technical can understand it,
     the title appears in PTB changelogs -->
#### Brief overview of PR changes/additions
Converts HTML to plain text in strings that are sent to screen readers.  Things like the error message in the script editor have things like bold in them, and any user defined item name with a less than sign in it such as a trigger named `hp < 200` should be escaped properly at this point thanks to earlier updates that were focused on proper displays.

#### Motivation for adding to Mudlet
This was a goal after #6795

#### Other info (issues closed, discussion etc)

atari2600tim committed:
Update mudlet.cpp

keneanung commented:
/refresh links

atari2600tim committed:
Merge branch 'development' into timStripHtmlSpeech

atari2600tim committed:
Merge branch 'development' into timStripHtmlSpeech

atari2600tim committed:
Merge branch 'development' into timStripHtmlSpeech

atari2600tim committed:
Merge branch 'development' into timStripHtmlSpeech

atari2600tim committed:
Merge branch 'Mudlet:development' into timStripHtmlSpeech

atari2600tim committed:
Merge branch 'Mudlet:development' into timStripHtmlSpeech

vadi2 commented:
@atari2600tim do you want to make this ready for review?

atari2600tim committed:
Merge branch 'Mudlet:development' into timStripHtmlSpeech

atari2600tim committed:
Merge branch 'development' into timStripHtmlSpeech

atari2600tim committed:
optional arg to announce func for plain text

atari2600tim committed:
Merge branch 'timStripHtmlSpeech' of https://github.com/atari2600tim/Mudlet into timStripHtmlSpeech

atari2600tim committed:
Update mudlet.cpp

atari2600tim committed:
Update mudlet.cpp

atari2600tim commented:
Game output is not html, so I added an optional arg for it to utilize to say not html, made it and the lua function both use it.  Seems decent on the setups I tried so far (linux dev build on Ubuntu I think called Orca, and the bot build I downloaded the 64 bit version and ran it on Windows 10 with NVDA).  I don't know enough about our process to say if that failing test is related.

vadi2 commented on src/mudlet.cpp:
```diff
@@ -5114,9 +5114,14 @@ bool mudlet::desktopInDarkMode()
+    if (isPlain){
+        mpAnnouncer->announce(text, processing);
+    } else {
+        QTextEdit convertor(text);
```
This seems expensive to create a visual widget that can display rich text every time you need to read some text... how about using a static `QTextDocument` instead that is created once?

vadi2 commented on src/mudlet.h:
```diff
@@ -325,7 +325,7 @@ class mudlet : public QMainWindow, public Ui::main_window
     void adjustMenuBarVisibility();
     void adjustToolBarVisibility();
-    void announce(const QString& text, const QString& processing = QString());
+    void announce(const QString& text, const QString& processing = QString(), bool isPlain = false);
```
Won't this affect the Lua announce() function as well?

atari2600tim committed:
Merge branch 'Mudlet:development' into timStripHtmlSpeech

atari2600tim committed:
change QTextEdit to QTextDocument

must have got example from web, certainly don't want a whole editor thing in this situation
also no need to hold in a variable

atari2600tim committed:
Merge branch 'development' into timStripHtmlSpeech

atari2600tim committed:
Update mudlet.cpp

atari2600tim committed:
Merge branch 'timStripHtmlSpeech' of https://github.com/atari2600tim/Mudlet into timStripHtmlSpeech

atari2600tim committed:
Merge branch 'Mudlet:development' into timStripHtmlSpeech

atari2600tim committed:
Merge branch 'Mudlet:development' into timStripHtmlSpeech

atari2600tim committed:
Merge branch 'Mudlet:development' into timStripHtmlSpeech

atari2600tim committed:
Merge branch 'development' into timStripHtmlSpeech

atari2600tim committed:
Merge branch 'Mudlet:development' into timStripHtmlSpeech

atari2600tim committed:
Merge branch 'development' into timStripHtmlSpeech

atari2600tim committed:
Merge branch 'development' into timStripHtmlSpeech

atari2600tim committed:
Merge branch 'development' into timStripHtmlSpeech

atari2600tim committed:
Merge branch 'Mudlet:development' into timStripHtmlSpeech

atari2600tim committed:
Merge branch 'Mudlet:development' into timStripHtmlSpeech

atari2600tim committed:
Merge branch 'Mudlet:development' into timStripHtmlSpeech

atari2600tim committed:
Merge branch 'Mudlet:development' into timStripHtmlSpeech

ZookaOnGit committed:
Merge branch 'development' into timStripHtmlSpeech

atari2600tim committed:
Merge branch 'development' into timStripHtmlSpeech

atari2600tim committed:
Merge branch 'development' into timStripHtmlSpeech

ZookaOnGit commented:
> @atari2600tim do you want to make this ready for review?

atari2600tim commented:
I'm not happy with it yet, does not do all that I had intended.  Am planning to get some progress tomorrow.

atari2600tim committed:
Merge branch 'development' into timStripHtmlSpeech

atari2600tim committed:
Merge branch 'development' into timStripHtmlSpeech

atari2600tim committed:
Update mudlet.cpp

atari2600tim commented:
Here is a video showing some parts that it changes, comparing current PTB on left and development build on the right.  I typed up what I was doing for the most part, but I did not highlight and point at the notes like I normally would, because doing so would have made screen reader react to me doing that.  Some parts it is not obvious that I pushed Control-3 to switch to Scripts view when I was already on Scripts view.  Hovering mouse over things was good already, but the reading upon switch or upon showing item with an error is what changes here.

https://github.com/user-attachments/assets/f56963f5-268f-4823-befa-64f637c7805a

atari2600tim committed:
Merge branch 'development' into timStripHtmlSpeech

vadi2 reviewed (approved):
Thank you!
